src/Interpreters/ExpressionActions.h

```diff
@@ -28,13 +31,138 @@ using ArrayJoinActionPtr = std::shared_ptr<ArrayJoinAction>;
+    /// For each node we need to know all it's ancestors that are short-circuit functions.
+    /// Also we need to know which arguments of this short-circuit functions are ancestors for the node
+    /// (we will store the set of indexes of arguments), because for some short-circuit function we shouldn't
+    /// enable lazy execution for nodes that are common descendants of different function arguments.
+    /// Example: if(cond, expr1(..., expr, ...), expr2(..., expr, ...))).
+    std::unordered_map<const ActionsDAG::Node *, std::unordered_set<size_t>> short_circuit_ancestors_info;
+};
+
+/// To determine whether a node could be executed lazily.
+/// Let's make it clear that in which situations a node could not be executed lazily.
+/// 1. it's not a function or alias node
+/// 2. it doesn't have any parent.
+/// 3. it is used as a output attribute of this ExpressionActions.
+/// 4. it's the first node of one of its parent, and the parent's
+///    enable_lazy_execution_for_first_argument = false.
+/// 5. it's a common descendant of different function arguments of short-circuit functions
+/// 6. it's parent could not be executed lazily and the parent is not a short-circuit function.
+class ExpressionShortCircuitExecuteController
+{
+public:
+    struct ProfileData
+    {
+        UInt64 sample_rows = 0;
+        UInt64 selected_rows = 0;
+        UInt64 eclapsed = 0;
+    };
+
+    enum ProfileType
+    {
+        NOT_PROFILE = 0,
+        PROFILE_ECLAPSED = 1,
+        PROFILE_SELECTIVITY = 2,
+    };
+
+    explicit ExpressionShortCircuitExecuteController(
+        const ActionsDAGPtr & actions_dag_,
+        const ExpressionActionsSettings & settings);
+
+    ExpressionShortCircuitExecuteController(const ExpressionShortCircuitExecuteController & b)
+        : actions_dag(b.actions_dag)
+        , short_circuit_function_evaluation(b.short_circuit_function_evaluation)
+        , enable_adaptive_reorder_arguments(b.enable_adaptive_reorder_arguments)
+        , sampled_rows(b.sampled_rows.load())
+        , finished_adaptive_reorder_arguements(b.finished_adaptive_reorder_arguements.load())
+        , short_circuit_infos(b.short_circuit_infos)
+        , has_reorderable_short_circuit_functions(b.has_reorderable_short_circuit_functions)
+    {}
+
+    ExpressionShortCircuitExecuteController & operator=(const ExpressionShortCircuitExecuteController & b)
+    {
+        actions_dag = b.actions_dag;
+        short_circuit_function_evaluation = b.short_circuit_function_evaluation;
+        enable_adaptive_reorder_arguments = b.enable_adaptive_reorder_arguments;
+        sampled_rows = b.sampled_rows.load();
+        finished_adaptive_reorder_arguements = b.finished_adaptive_reorder_arguements.load();
+        short_circuit_infos = b.short_circuit_infos;
+        has_reorderable_short_circuit_functions = b.has_reorderable_short_circuit_functions;
+        return *this;
+    }
+
+    bool couldLazyExecuted(const ActionsDAG::Node * node);
```

Comment:
Please, add comments for the methods below what they are supposed to do and return